Website/Providers/DataProviders/SqlDataProvider/09.02.00.SqlDataProvider

```diff
@@ -492,16 +492,87 @@ BEGIN
                                                 INNER JOIN {databaseOwner}{objectQualifier}UserProfile AS P ON U.UserID = P.UserID
-                                                LEFT OUTER JOIN {databaseOwner}{objectQualifier}ProfilePropertyDefinition AS Prop ON 
+                                                LEFT OUTER JOIN {databaseOwner}{objectQualifier}ProfilePropertyDefinition AS Prop ON
                                                 (Prop.PropertyDefinitionID = P.PropertyDefinitionID and Prop.Deleted = 0 and Prop.PortalID = ' + CONVERT(nvarchar(20), @PortalID) + ')
                                             WHERE U.UserId IN (SELECT UserId FROM @TempUsers) AND (U.PortalId = ' + CONVERT(nvarchar(20), @PortalID) + ' OR U.PortalId IS NULL)
                                             )' +
-                @pivotSql + @sortSql + ' DROP TABLE #MatchingUsers'           
+                @pivotSql + @sortSql + ' DROP TABLE #MatchingUsers'
 
         EXEC(@sql)
 END
 GO
 
+IF EXISTS (SELECT * FROM dbo.sysobjects WHERE id = object_id(N'{databaseOwner}[{objectQualifier}AddPasswordHistory]') AND OBJECTPROPERTY(id, N'IsPROCEDURE') = 1)
+	DROP PROCEDURE {databaseOwner}{objectQualifier}AddPasswordHistory
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}AddPasswordHistory]
+	@UserId             int,
+	@Password           nvarchar(128),
+	@PasswordSalt       nvarchar(128),
+	@PasswordsRetained  int,
+	@DaysRetained       int,
+	@CreatedByUserID    int
+AS
+
+	BEGIN
+
+	INSERT INTO {databaseOwner}{objectQualifier}PasswordHistory (
+		UserId,
+		[Password],
+		PasswordSalt,
+		CreatedByUserID,
+		CreatedOnDate,
+		LastModifiedByUserID,
+		LastModifiedOnDate
+	)
+	VALUES (
+		@UserId,
+		@Password,
+		@PasswordSalt,
+		@CreatedByUserID,
+		GETDATE(),
+		@CreatedByUserID,
+		GETDATE()
+	)
+
+	DELETE FROM {databaseOwner}{objectQualifier}PasswordHistory
+	WHERE UserID=@UserId
+	  AND PasswordHistoryID NOT IN (
+		SELECT TOP (@PasswordsRetained) PasswordHistoryID
+		FROM {databaseOwner}{objectQualifier}PasswordHistory
+		WHERE UserID=@UserId
+		ORDER BY CreatedOnDate DESC
+		)
+	  AND DATEDIFF(day, CreatedOnDate, GETDATE()) > @DaysRetained
+
+	END
+GO
+
+IF EXISTS (SELECT * FROM dbo.sysobjects WHERE id = object_id(N'{databaseOwner}[{objectQualifier}GetPasswordHistory]') AND OBJECTPROPERTY(id, N'IsPROCEDURE') = 1)
+	DROP PROCEDURE {databaseOwner}{objectQualifier}GetPasswordHistory
+GO
+
+CREATE PROCEDURE {databaseOwner}[{objectQualifier}GetPasswordHistory]
+	@UserID             int,
+	@PasswordsRetained  int,
+	@DaysRetained       int
+AS
+	DELETE FROM {databaseOwner}{objectQualifier}PasswordHistory
+	WHERE UserID=@UserId
+	  AND PasswordHistoryID NOT IN (
+		SELECT TOP (@PasswordsRetained) PasswordHistoryID
+		FROM {databaseOwner}{objectQualifier}PasswordHistory
+		WHERE UserID=@UserId
+		ORDER BY CreatedOnDate DESC
+		)
+	  AND DATEDIFF(day, CreatedOnDate, GETDATE()) > @DaysRetained;
```

Comment:
There is no need for this in this SPROC. This is for returning the old password. This is already achieved when you are adding a new password. The select statement can return top (xx) passwords withing a specific period of time.

Reply:
The purpose of this change is that you can specify a minimum number of days to retain passwords, in addition to a minimum number of passwords to retain, so that you can't just change your password many times in one day in order to get around the restrictions of password retention.  So, `AddPasswordHistory` may not remove your old password today, but that old password may no longer be relevant tomorrow, which is why this change ensures that the list of password history respects the retention rules before retrieving the passwords.  If it didn't do this, a password which could be reused today would still be in the list.  

Let me know if there's something I'm missing, or if there's a different solution you'd prefer to see.

Reply:
The purpose of this change is that you can specify a minimum number of days to retain passwords, in addition to a minimum number of passwords to retain, so that you can't just change your password many times in one day in order to get around the restrictions of password retention. So, `AddPasswordHistory` may not remove your old password today, but that old password may no longer be relevant tomorrow, which is why this change ensures that the list of password history respects the retention rules before retrieving the passwords. If it didn't do this, a password which could be reused today would still be in the list.

Let me know if there's something I'm missing, or if there's a different solution you'd prefer to see.

(sorry if you're getting these comments twice, I've not gotten the hang of GitHub reviews)